irc-proto/src/prefix.rs

```diff
@@ -0,0 +1,204 @@
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Prefix {
+    /// servername, e.g. collins.mozilla.org
+    ServerName(String),
+    /// nickname [ ["!" username] "@" hostname ]
+    /// i.e. Nickname(nickname, username, hostname)
+    /// Any of the strings may be ""
+    Nickname(String, String, String),
+}
+
+impl Prefix {
+    /// Creates a prefix by parsing a string.
+    ///
+    /// # Example
+    /// ```
+    /// # extern crate irc;
+    /// # use irc::client::prelude::*;
+    /// # fn main() {
+    /// Prefix::new_from_str("nickname!username@hostname");
+    /// Prefix::new_from_str("example.com");
+    /// # }
+    /// ```
+    pub fn new_from_str(s: &str) -> Prefix {
+        #[derive(Copy, Clone, Eq, PartialEq)]
+        enum Active {
+            Name = 0,
+            User = 1,
+            Host = 2,
+        }
+
+        let mut name = String::new();
+        let mut user = String::new();
+        let mut host = String::new();
+        let mut active = Active::Name;
+
+        for c in s.chars() {
+            match c {
+                // We consider the '.' to be a ServerName except if a ! has already
+                // been encountered.
+                '.' if active == Active::Name => return Prefix::ServerName(s.to_owned()),
+
+                '!' if active == Active::Name => {
+                    active = Active::User;
+                },
+
+                // The '@' is not special until we've started the username
+                // portion
+                '@' if active == Active::User => {
+                    active = Active::Host;
+                },
+
+                _ => {
+                    // Push onto the latest buffer
+                    match active {
+                        Active::Name => &mut name,
+                        Active::User => &mut user,
+                        Active::Host => &mut host,
+                    }.push(c)
+                }
+            }
+        }
+
+        Prefix::Nickname(name, user, host)
+    }
+}
+
+/// This implementation never returns an error and is isomorphic with `Display`.
+impl FromStr for Prefix {
+    type Err = string::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Prefix::new_from_str(s))
+    }
+}
+
+/// This is isomorphic with `FromStr`
+impl fmt::Display for Prefix {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Prefix::ServerName(name) => write!(f, "{}", name),
+            Prefix::Nickname(name, user, host) => match (&name[..], &user[..], &host[..]) {
+                ("", "", "") => write!(f, ""),
+                (name, "", "") => write!(f, "{}", name),
+                (name, user, "") => write!(f, "{}!{}", name, user),
+                // This case shouldn't happen normally, but user!@host is invalid, so we
+                // format it without the host
+                (name, "", _host) => write!(f, "{}", name),
```

Comment:
`nickname!@hostname` is invalid, but `nickname@hostname` is valid, and wouldn't the latter also result in this match arm being used?